core/api/views/cp_reports.py

```diff
@@ -118,6 +134,125 @@ def customize_errors(self, error_dict):
+        for cp_record in instance.cprecords.all():
+            cp_record_ar = self._get_archive_data(
+                CPRecordArchive,
+                cp_record,
+                {"country_programme_report_id": cp_report_ar.id},
+            )
+            cp_record_ar.save()
+            for cp_usage in cp_record.record_usages.all():
+                cp_usages.append(
+                    self._get_archive_data(
+                        CPUsageArchive,
+                        cp_usage,
+                        {"country_programme_record_id": cp_record_ar.id},
+                    )
+                )
+        CPUsageArchive.objects.bulk_create(cp_usages, batch_size=1000)
+
+        # archive cp_prices
+        cp_prices = []
+        for cp_price in instance.prices.all():
+            cp_prices.append(
+                self._get_archive_data(
+                    CPPricesArchive,
+                    cp_price,
+                    {"country_programme_report_id": cp_report_ar.id},
+                )
+            )
+        CPPricesArchive.objects.bulk_create(cp_prices, batch_size=1000)
+
+        # archive cp_generation
+        cp_generation = []
+        for cp_gen in instance.cpgenerations.all():
+            cp_generation.append(
+                self._get_archive_data(
+                    CPGenerationArchive,
+                    cp_gen,
+                    {"country_programme_report_id": cp_report_ar.id},
+                )
+            )
+        CPGenerationArchive.objects.bulk_create(cp_generation, batch_size=1000)
+
+        # archive cp_emission
+        cp_emission = []
+        for cp_em in instance.cpemissions.all():
+            cp_emission.append(
+                self._get_archive_data(
+                    CPEmissionArchive,
+                    cp_em,
+                    {"country_programme_report_id": cp_report_ar.id},
+                )
+            )
+        CPEmissionArchive.objects.bulk_create(cp_emission, batch_size=1000)
+
+    @transaction.atomic
+    def put(self, request, *args, **kwargs):
+        instance = self.get_object()
+        if instance.year <= IMPORT_DB_MAX_YEAR:
+            return Response(
+                {
+                    "error": "Cannot update country programme report with year <="
+                    f" {IMPORT_DB_MAX_YEAR}"
+                },
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        serializer = CPReportCreateSerializer(data=request.data)
+        if not serializer.is_valid():
+            custom_errors = self.customize_errors(serializer.errors)
+            return Response(custom_errors, status=status.HTTP_400_BAD_REQUEST)
+        self.perform_create(serializer)
+        # update version number
+        new_instance = serializer.instance
+        # increment version number if the status is FINAL
+        new_instance.version = instance.version + int(
+            instance.status == CPReport.CPReportStatus.FINAL
+        )
```

Comment:
Since we are incrementing this in python, we probably need to use [select_for_update](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#select-for-update) to ensure the relevant row is locked. To avoid the very unlikely case of a race condition here.